tasks/kernel_matrix_testing/compiler.py

```diff
@@ -1,14 +1,23 @@
+from __future__ import annotations
+
 import sys
+from typing import TYPE_CHECKING, Optional, cast
+
+from invoke.context import Context
+from invoke.runners import Result
 
 from tasks.kernel_matrix_testing.tool import info
 
+if TYPE_CHECKING:
+    from tasks.kernel_matrix_testing.types import PathOrStr
+
 
-def compiler_built(ctx):
+def compiler_built(ctx: Context):
     res = ctx.run("docker images kmt:compile | grep -v REPOSITORY | grep kmt", warn=True)
-    return res.ok
+    return res is not None and res.ok
 
 
-def docker_exec(ctx, cmd, user="compiler", verbose=True, run_dir=None):
+def docker_exec(ctx: Context, cmd: str, user="compiler", verbose=True, run_dir: Optional[PathOrStr] = None):
```

Comment:
does `PathOrStr` cause problems if not type checking, since the import isn't present?

Reply:
No, the file starts with `from __future__ import annotations` (which is available [from Python 3.7 onwards](https://peps.python.org/pep-0563/)) so type annotations are not evaluated during runtime.